torchmetrics/detection/map.py

```diff
@@ -279,20 +291,339 @@ def update(self, preds: List[Dict[str, Tensor]], target: List[Dict[str, Tensor]]
+                Class Id of the supplied groundtruth and detection labels.
+            area_range:
+                List of lower and upper bounding box area threshold.
+            max_det:
+                Maximum number of evaluated detection bounding boxes.
+            ious:
+                IoU reults for image and class.
+        """
+        gt = self.groundtruth_boxes[id]
+        dt = self.detection_boxes[id]
+        gt_lbl_mask = self.groundtruth_labels[id] == class_id
+        dt_lbl_mask = self.detection_labels[id] == class_id
+        if len(dt_lbl_mask) == 0 or len(dt_lbl_mask) == 0:
+            return None
+        gt = gt[gt_lbl_mask]
+        dt = dt[dt_lbl_mask]
+        if len(gt) == 0 and len(dt) == 0:
+            return None
+
+        areas = box_area(gt)
+        ignore_area = (areas < area_range[0]) | (areas > area_range[1])
+
+        # sort dt highest score first, sort gt ignore last
+        ignore_area_sorted, gtind = torch.sort(ignore_area)
+        gt = gt[gtind]
+        scores = self.detection_scores[id]
+        scores_filtered = scores[dt_lbl_mask]
+        scores_sorted, dtind = torch.sort(scores_filtered, descending=True)
+        dt = dt[dtind]
+        if len(dt) > max_det:
+            dt = dt[:max_det]
+        # load computed ious
+        ious = ious[id, class_id][:, gtind] if len(ious[id, class_id]) > 0 else ious[id, class_id]
+
+        T = len(self.iou_thresholds)
+        G = len(gt)
+        D = len(dt)
+        gt_matches = torch.zeros((T, G), dtype=torch.bool)
+        dt_matches = torch.zeros((T, D), dtype=torch.bool)
+        gt_ignore = ignore_area_sorted
+        dt_ignore = torch.zeros((T, D), dtype=torch.bool)
+        if len(ious) > 0:
+            for tind, t in enumerate(self.iou_thresholds):
+                for d in range(D):
+                    # information about best match so far (m=-1 -> unmatched)
+                    iou = min([t, 1 - 1e-10])
+                    m = -1
+                    for g in range(G):
+                        # if this gt already matched, and not a crowd, continue
+                        if gt_matches[tind, g] > 0:
+                            continue
+                        # if dt matched to reg gt, and on ignore gt, stop
+                        if m > -1 and not gt_ignore[m] and gt_ignore[g]:
+                            break
+                        # continue to next gt unless better match made
+                        if ious[d, g] < iou:
+                            continue
+                        # if match successful and best so far, store appropriately
+                        iou = ious[d, g]
+                        m = g
+                    # if match made store id of match for both dt and gt
+                    if m == -1:
+                        continue
+
+                    dt_ignore[tind, d] = gt_ignore[m]
+                    dt_matches[tind, d] = True
+                    gt_matches[tind, m] = True
```

Comment:
it may make code easier to read if we move this body to a separate private function/method...